File: CompetitiveProgramming/BinaryTrees/verticalOrderTraversalOfBinaryTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    
    
    solution = {}
    
    maxHorizontalDistance = 0 
    
    minHorizontalDistance = 0 
    
    def parseTree(self,node,level,horizontalDistance):
        
        if(node == None):
            return 
        
       
        
        if(horizontalDistance not in self.solution):
            
            self.solution[horizontalDistance] = [(level,node.val)]
            
            if(self.maxHorizontalDistance < horizontalDistance):
                self.maxHorizontalDistance = horizontalDistance
                
            if(self.minHorizontalDistance > horizontalDistance):
                self.minHorizontalDistance = horizontalDistance
            
        else:
            
            arr =  self.solution[horizontalDistance]
            
            
            for i in range(len(arr)):
                
                #if levels are equal
                if(level == arr[i][0]):
                    
                    val = arr[i][1]
                    
                    if(val < node.val):
                        pass
                    else:
                        arr.insert(i,(level,node.val))
                        break

                elif(level < arr[i][0]):
                    
                    arr.insert(i,(level,node.val))
                    break
                    
                
                if(i == len(arr)-1):

                    arr.append((level,node.val))
        
        self.parseTree(node.left,level+1,horizontalDistance-1)
        
        self.parseTree(node.right,level+1,horizontalDistance + 1)
                    
        return
                    
        
            
            

    def verticalTraversal(self, root):
        
        self.solution = {}
        
        self.maxHorizontalDistance = 0 
        
        self.minHorizontalDistance = 0 
        
        self.parseTree(root,0,0)
        
        result = []
        
        
        for i in range(self.minHorizontalDistance, self.maxHorizontalDistance + 1):
            
            arr = self.solution[i]
            
            levelele = []
            
            for j in range(len(arr)):
                levelele.append(arr[j][1])
            
            result.append(levelele)
            
        return result 
            
            
        
        

File: CompetitiveProgramming/BinaryTrees/verticalOrderTraversalOfBinaryTree/verticalOrderTraversalOfBinaryTree.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNaode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include<vector> 

class Solution {
    
private:
    int minHorizontalDistance;
    
private:
    int maxHorizontalDistance;


private:
    void parseTree(TreeNode* node, 
                   int level, 
                   int horizontalDistance, 
                   map<int,vector<vector<int>>> &levelNodes) {
        
        
        if(node == nullptr){
            return;
        }
        

        
        if(horizontalDistance > maxHorizontalDistance){
            maxHorizontalDistance = horizontalDistance;
        }
        
        if(horizontalDistance < minHorizontalDistance){
            minHorizontalDistance = horizontalDistance;
        }
        
        vector<int> nodeVector {level,node->val};
        
        if(levelNodes.find(horizontalDistance) == levelNodes.end()){
            
            levelNodes[horizontalDistance].push_back(nodeVector);
        }
        else{
            vector<vector<int>> arr = levelNodes[horizontalDistance];
            
            for(int i = 0; i< arr.size(); i++){
                
                if(level == arr[i][0]){
                    if(node->val <= arr[i][1]){
                        levelNodes[horizontalDistance].insert(levelNodes[horizontalDistance].begin()+i ,nodeVector);
                        break;
                    }
                  
                }
                
                if(level < arr[i][0]){
                    
                    levelNodes[horizontalDistance].insert(levelNodes[horizontalDistance].begin()+i , nodeVector);
                    break;
                        
                }
                
                if(i == arr.size()-1){
                    levelNodes[horizontalDistance].push_back(nodeVector);
                }
                    
                
            }
        }
        
        
        
        parseTree(node->left, level + 1, horizontalDistance -1 , levelNodes);
        parseTree(node->right, level +1 , horizontalDistance +1 ,levelNodes);
        
        return;
        
    }
    
    
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        
        minHorizontalDistance = 0;
        maxHorizontalDistance = 0;
        
        map<int,vector<vector<int>>> levelNodes;
        
        parseTree(root, 0, 0 , levelNodes);
            
        vector<vector<int>> results;
        
        for(int i = minHorizontalDistance ; i <= maxHorizontalDistance ; i++){
            
            vector<vector<int>> arr = levelNodes[i];
                
            vector<int> result;
                
            for(int j = 0 ;j < arr.size(); j++){
                
                result.push_back(arr[j][1]);
                

            }
            
            results.push_back(result);
        }

        
        return results;
        
        
    }
};